src/rgw/rgw_sal.h

```diff
@@ -468,10 +482,15 @@ class User {
 
     /** Clone a copy of this user.  Used when modification is necessary of the copy */
     virtual std::unique_ptr<User> clone() = 0;
-    /** List the buckets owned by a user */
+
+    /// List the user's buckets, copying entries into the provided \ref storage
+    /// sorted by bucket name.
     virtual int list_buckets(const DoutPrefixProvider* dpp,
-			     const std::string& marker, const std::string& end_marker,
-			     uint64_t max, bool need_stats, BucketList& buckets,
+			     const std::string& marker,
+			     const std::string& end_marker,
+			     bool need_stats,
+			     std::span<RGWBucketEnt> storage,
```

Comment:
It's not really obvious what role `storage` is playing. Is it a cache to store more RGWBucketEnt's than are intended to be returned at each call to list_buckets? This needs to be nicely documented here, so future callers don't have to trace through the code to figure what's happening. Would a name such as `cache` or `temp_storage` help clarify, I wonder.

Reply:
> It's not really obvious what role storage is playing

thanks @ivancich. this feedback from you and @dang has convinced me that the interface isn't as nice as i'd thought

i'm mainly trying to get away from functions that have to allocate the results for you. so this required the caller to pass in preallocated storage; the `std::span<RGWBucketEnt> storage` argument would allow the caller to pass in an array on the stack if they wanted to avoid the heap altogether

but in this specific case, `rgw_list_buckets_max_chunk` controls the chunk size of each listing, so none of the callers were able to use a static array anyway. i'll change `BucketList::buckets` from a span to a vector so that the results get allocated as before. at least this vector can be reused over multiple calls to `list_buckets()`